### PR TITLE
Allow the head element to be optional

### LINKED ISSUE #241: Do not require `<head>` and `<body>` elements for transformers
We really shouldn't be requiring `<head>` and `<body>` elements [in transformers/optimizers] as those are not required in HTML. What we should be doing in a case like this is if there is not a head, find the first non-head element (so anything that's not one of these: https://htmlhead.dev/#elements) and inject before that element. The browser will parse it as inside of the head. We could create a utility to make these easier for transformers to do.

_Originally posted by @matthewp in https://github.com/snowpackjs/astro/pull/231#discussion_r638036754_

### PR DESCRIPTION
**Pull request by matthewp**
Fixes #241 

## Changes

This makes it possible to omit the head element but still have styles/scripts that we inject put into the head section (by tracking when we encounter a non-head element).

## Testing

Test added.

## Docs

This should work already, as per HTML spec.

### CODE REVIEW

**Comment by matthewp** [comment]
This has a technical issue in that we currently do not differentiate between page components and regular astro components in compilation. This means that in practice the way we compile page components in a special way is by finding the head element.

File: packages/astro/src/compiler/transform/head.ts [hunk 1]
```diff
@@ -1,41 +1,114 @@
...
+
+      /**
+       * Injects an expression that adds link tags for provided css.
+       * Turns into:
+       * ```
+       * { Astro.css.map(css => (
+       *  <link rel="stylesheet" href={css}>
+       * ))}
+       * ```
+       */
```
Comment (natemoo-re) [inline packages/astro/src/compiler/transform/head.ts:51]: Woah. That's super cool. Nice solution!

File: packages/astro/src/compiler/transform/head.ts [hunk 2]
```diff
@@ -79,8 +152,26 @@ export default function (opts: TransformOptions): Transformer {
...
-      head.children = head.children ?? [];
-      head.children.push(...children);
+
+      const conditionalNode = {
+        start: 0,
+        end: 0,
+        type: 'Expression',
+        codeChunks: [
+          'Astro.isPage ? (',
+          ') : null'
```
Comment (natemoo-re) [inline packages/astro/src/compiler/transform/head.ts:162]: 🤯

File: packages/astro/src/runtime.ts [hunk 3]
```diff
@@ -228,17 +228,9 @@ async function load(config: RuntimeConfig, rawPathname: string | undefined): Pro
...
 
-    // inject styles
-    // TODO: handle this in compiler
-    const styleTags = Array.isArray(mod.css) && mod.css.length ? mod.css.reduce((markup, href) => `${markup}\n<link rel="stylesheet" type="text/css" href="${href}" />`, '') : ``;
-    if (html.indexOf('</head>') !== -1) {
-      html = html.replace('</head>', `${styleTags}</head>`);
-    } else {
-      html = styleTags + html;
-    }
-
```
Comment (natemoo-re) [inline packages/astro/src/runtime.ts:241]: Nice!

👋 ~TODO~

**Review by natemoo-re — APPROVED** [review]
This looks awesome, great work!

**Comment by matthewp** [comment]
Note to self: Add a changeset before merging.

**Pull request merged**